File: apps/api/plane/app/serializers/stage_review.py
"""阶段评审实例的接口契约。

三层输出：列表行（一个阶段的全部评审，扁平返回、由前端按产品分组并把评审活动缩进到
所属评审下）、详情（带描述、工作指引、O 阶段分组、附件与评论计数），以及活动与评论。
输入按动作拆成几个小序列化器，形状校验留在这里，跨表规则（父评审是否同产品、结论是否
齐备）在 ``utils/stage_review.py``。

**评审与评审活动共用同一套序列化器** —— 两者字段几乎一样，只有层级不同（产品决策）。
前端也共用同一个抽屉，所以这里刻意不按 kind 分叉。
"""

from rest_framework import serializers

from plane.app.serializers.data_dictionary import DataDictionaryItemLiteSerializer
from plane.app.serializers.user import UserLiteSerializer
from plane.db.models import (
    ProjectMember,
    StageReview,
    StageReviewActivity,
    StageReviewComment,
    StageReviewKind,
    StageReviewResult,
    User,
)
from plane.db.models.stage_review import (
    COMPONENT_VERSION_FIELDS,
    FINISHED_GOODS_FIELDS,
    ProductionMode,
    ShipmentAssessment,
)

from .base import BaseSerializer


class StageReviewProductSerializer(serializers.Serializer):
    """列表分组头要的四个字段。手写而不是复用 ProductSerializer：那个太重。"""

    id = serializers.UUIDField(read_only=True)
    name = serializers.CharField(read_only=True)
    code = serializers.CharField(read_only=True)
    identifier = serializers.CharField(read_only=True)


class StageReviewProjectSerializer(serializers.Serializer):
    """产品页按项目分组 / 项目列 / 抽屉面包屑要的字段。

    不复用 ProjectLiteSerializer：那个带 cover_image_url，会逐行触碰封面资产外键。
    """

    id = serializers.UUIDField(read_only=True)
    name = serializers.CharField(read_only=True)
    identifier = serializers.CharField(read_only=True)
    logo_props = serializers.JSONField(read_only=True)


class StageReviewListSerializer(BaseSerializer):
    """列表行。附件数由 view 的 queryset annotate 出来，不在这里反查。

    ``is_manual`` 就是「模板为空」—— 手工新建的评审不进裁剪表，列表要能一眼认出来。
    ``project_detail`` 给产品页用（一个产品横跨多个项目）；queryset 都已 select_related project。
    """

    project_detail = StageReviewProjectSerializer(source="project", read_only=True)
    product_id = serializers.UUIDField(read_only=True)
    product_detail = StageReviewProductSerializer(source="product", read_only=True)
    stage_id = serializers.UUIDField(read_only=True)
    parent_id = serializers.UUIDField(read_only=True)
    template_id = serializers.UUIDField(read_only=True)
    # 来源裁剪表：同一产品的同一评审在多张表里都保留时会各生成一条，列表靠这个区分。
    # 反向关系，由 view annotate 出来，不在这里逐行反查；手工新建的评审为空。
    tailoring_id = serializers.UUIDField(read_only=True, default=None)
    tailoring_title = serializers.CharField(read_only=True, default=None)
    leader_detail = UserLiteSerializer(source="leader", read_only=True)
    auditor_detail = UserLiteSerializer(source="auditor", read_only=True)
    attachment_count = serializers.IntegerField(read_only=True, default=0)
    comment_count = serializers.IntegerField(read_only=True, default=0)
    is_manual = serializers.SerializerMethodField()

    class Meta:
        model = StageReview
        fields = [
            "id",
            "project_id",
            "project_detail",
            "workspace_id",
            "product_id",
            "product_detail",
            "stage_id",
            "parent_id",
            "template_id",
            "tailoring_id",
            "tailoring_title",
            "kind",
            "title",
            "status",
            "result",
            "leader_id",
            "leader_detail",
            "auditor_id",
            "auditor_detail",
            "start_date",
            "end_date",
            "attachment_count",
            "comment_count",
            "is_manual",
            "sort_order",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields

    def get_is_manual(self, obj) -> bool:
        return obj.template_id is None


class StageReviewDetailSerializer(StageReviewListSerializer):
    """抽屉里要的全部字段。

    成品与组件版本按分组吐 json（模型上的两个 property），只有「O阶段评审」有值；
    生产方式与出货评估两种 O 阶段类型都有。前端按 ``kind`` 决定渲不渲这两块。
    """

    stage_detail = DataDictionaryItemLiteSerializer(source="stage", read_only=True)
    # 评审活动的面包屑要带一级「所属评审」，只要标题，不展开整个父对象
    parent_title = serializers.CharField(
        source="parent.title", read_only=True, allow_null=True, default=None
    )
    finished_goods = serializers.DictField(read_only=True)
    component_versions = serializers.DictField(read_only=True)

    class Meta(StageReviewListSerializer.Meta):
        fields = StageReviewListSerializer.Meta.fields + [
            "stage_detail",
            "parent_title",
            "description_html",
            "work_instruction",
            "conditional_reason",
            "initiator_role",
            "leader_role",
            "auditor_role",
            "production_mode",
            "shipment_assessment",
            "finished_goods",
            "component_versions",
            "created_by",
        ]
        read_only_fields = fields


class StageReviewActivitySerializer(BaseSerializer):
    actor_detail = UserLiteSerializer(source="actor", read_only=True)

    class Meta:
        model = StageReviewActivity
        fields = [
            "id",
            "stage_review",
            "actor",
            "actor_detail",
            "verb",
            "field",
            "old_value",
            "new_value",
            "comment",
            "stage_review_comment",
            "old_identifier",
            "new_identifier",
            "epoch",
            "extra",
            "created_at",
        ]
        read_only_fields = fields


class StageReviewCommentSerializer(BaseSerializer):
    actor_detail = UserLiteSerializer(source="actor", read_only=True)

    class Meta:
        model = StageReviewComment
        fields = [
            "id",
            "workspace",
            "project",
            "stage_review",
            "actor",
            "actor_detail",
            "comment_stripped",
            "comment_json",
            "comment_html",
            "parent",
            "edited_at",
            "created_at",
            "updated_at",
            "created_by",
            "updated_by",
        ]
        read_only_fields = [
            "workspace",
            "project",
            "stage_review",
            "actor",
            "comment_stripped",
            "edited_at",
            "created_by",
            "updated_by",
            "created_at",
            "updated_at",
        ]


# --- 输入 -----------------------------------------------------------------


class StageReviewCreateSerializer(serializers.Serializer):
    """手工新建。``template`` 恒为空，由 utils 负责，这里只校验形状。"""

    product_id = serializers.UUIDField()
    stage_id = serializers.UUIDField()
    kind = serializers.ChoiceField(choices=StageReviewKind.choices)
    parent_id = serializers.UUIDField(required=False, allow_null=True)
    title = serializers.CharField(max_length=255)
    description_html = serializers.CharField(
        required=False, allow_blank=True, allow_null=True, default=""
    )
    work_instruction = serializers.CharField(
        required=False, allow_blank=True, allow_null=True, default=""
    )
    leader_id = serializers.UUIDField(required=False, allow_null=True)
    auditor_id = serializers.UUIDField(required=False, allow_null=True)
    start_date = serializers.DateField(required=False, allow_null=True)
    end_date = serializers.DateField(required=False, allow_null=True)


class StageReviewUpdateSerializer(serializers.ModelSerializer):
    """详情页里能改的字段。

    **``status`` 与 ``result`` 不在其中** —— 状态只能由动作按顺序推进，结论只能在
    「提交审核」那一刻写入，开个 PATCH 口子等于把状态机架空。
    """

    class Meta:
        model = StageReview
        fields = [
            "title",
            "description_html",
            "work_instruction",
            "leader",
            "auditor",
            "start_date",
            "end_date",
            *FINISHED_GOODS_FIELDS,
            *COMPONENT_VERSION_FIELDS,
        ]


#: 批量改属性一次最多改多少条。列表一次取全，一个项目一两百条，留足余量
STAGE_REVIEW_BULK_LIMIT = 500

#: 批量能改的属性。只放「一批评审能共用同一个值」的字段 —— 标题、描述、O 阶段成品信息
#: 各条不同，状态与结论只能由本人推进（见 ``StageReviewUpdateSerializer``）
STAGE_REVIEW_BULK_FIELDS = ("leader", "auditor", "start_date", "end_date")


class StageReviewBulkUpdateSerializer(serializers.Serializer):
    """列表勾选后批量改属性。

    属性字段「出现即修改」，不出现就保持不变；``leader`` / ``auditor`` 传 null 表示清空。
    负责人 / 审核者必须是本项目的活跃成员（``context["project_id"]``）—— 单条抽屉里按产品
    角色筛候选，但一批评审的角色名各不相同，批量只能退到项目成员这一层。
    """

    review_ids = serializers.ListField(
        child=serializers.UUIDField(),
        allow_empty=False,
        max_length=STAGE_REVIEW_BULK_LIMIT,
    )
    leader = serializers.PrimaryKeyRelatedField(
        queryset=User.objects.filter(is_active=True), required=False, allow_null=True
    )
    auditor = serializers.PrimaryKeyRelatedField(
        queryset=User.objects.filter(is_active=True), required=False, allow_null=True
    )
    start_date = serializers.DateField(required=False, allow_null=True)
    end_date = serializers.DateField(required=False, allow_null=True)

    def _validate_member(self, user):
        if user is None:
            return user
        if not ProjectMember.objects.filter(
            project_id=self.context.get("project_id"),
            member_id=user.id,
            is_active=True,
        ).exists():
            raise serializers.ValidationError("该成员不在本项目中")
        return user

    def validate_leader(self, value):
        return self._validate_member(value)

    def validate_auditor(self, value):
        return self._validate_member(value)

    def validate(self, attrs):
        if not any(field in attrs for field in STAGE_REVIEW_BULK_FIELDS):
            raise serializers.ValidationError("没有要修改的属性")
        start, end = attrs.get("start_date"), attrs.get("end_date")
        if start and end and end < start:
            raise serializers.ValidationError({"end_date": "结束日期不能早于开始日期"})
        return attrs


class StageReviewRollbackSerializer(serializers.Serializer):
    """退回上一步时的理由。非空判断同样下沉到 utils，这里只管类型。"""

    reason = serializers.CharField(allow_blank=True, default="")


class StageReviewApproveSerializer(serializers.Serializer):
    """审核通过时的审核意见，选填。只进轨迹的 extra，不落评审字段。"""

    approval_comment = serializers.CharField(
        required=False, allow_blank=True, default=""
    )


class StageReviewSubmitSerializer(serializers.Serializer):
    """提交审核时的结论。必填与否按 kind 判，规则在 utils 里。"""

    result = serializers.ChoiceField(choices=StageReviewResult.choices)
    conditional_reason = serializers.CharField(
        required=False, allow_blank=True, default=""
    )
    production_mode = serializers.ChoiceField(
        choices=ProductionMode.choices, required=False, allow_blank=True, default=""
    )
    shipment_assessment = serializers.ChoiceField(
        choices=ShipmentAssessment.choices,
        required=False,
        allow_blank=True,
        default="",
    )

File: packages/types/src/stage-review.ts
import type { TLogoProps } from "./common";
import type { TDataDictionaryItemLite } from "./data-dictionary";
import type { EStageReviewKind } from "./stage-review-template";
import type { IUserLite } from "./users";

/**
 * 阶段评审实例：裁剪表签批生效后，勾中的格子变成的一条条真实评审。
 *
 * **评审与评审活动是同一张表的两行**，只有层级不同（`kind` + `parent_id`），列表与
 * 详情抽屉共用同一套结构；父评审不汇总子活动的结论，各自独立评审。
 *
 * 状态只能顺着走：未评审 → 评审中 → 审核中 → 已评审，没有「直接改成某个状态」的写
 * 入口，所以前端也不该出现状态下拉框。「提交审核」的落点由结论决定：不通过留在评审中，
 * 免审直接已评审，通过 / 条件通过进审核中。
 */

export enum EStageReviewStatus {
  NOT_STARTED = "not_started",
  IN_REVIEW = "in_review",
  IN_APPROVAL = "in_approval",
  COMPLETED = "completed",
}

/** 四步流程条的顺序，也是「下一步 / 上一步」的唯一依据 */
export const STAGE_REVIEW_STATUS_ORDER: EStageReviewStatus[] = [
  EStageReviewStatus.NOT_STARTED,
  EStageReviewStatus.IN_REVIEW,
  EStageReviewStatus.IN_APPROVAL,
  EStageReviewStatus.COMPLETED,
];

export enum EStageReviewResult {
  PASSED = "passed",
  REJECTED = "rejected",
  WAIVED = "waived",
  CONDITIONAL = "conditional",
}

/** 生产方式 / 出货评估只在 O 阶段的两种类型上出现 */
export type TProductionMode = "normal" | "risk";
export type TShipmentAssessment = "normal" | "refresh" | "risk";

export const PRODUCTION_MODES: TProductionMode[] = ["normal", "risk"];
export const SHIPMENT_ASSESSMENTS: TShipmentAssessment[] = ["normal", "refresh", "risk"];

/** 列表按产品分组，分组头要的就是这四个字段 */
export type TStageReviewProduct = {
  id: string;
  name: string;
  code: string;
  identifier: string;
};

/** 所属项目：产品页横跨多个项目，按项目分组 / 项目列 / 抽屉面包屑都靠它 */
export type TStageReviewProject = {
  id: string;
  name: string;
  identifier: string;
  logo_props: TLogoProps;
};

export type TStageReview = {
  id: string;
  project_id: string;
  project_detail: TStageReviewProject | null;
  workspace_id: string;
  product_id: string;
  product_detail: TStageReviewProduct | null;
  stage_id: string;
  /** 评审活动指向它所属的评审；顶层评审为 null。前端靠它把列表折成两层 */
  parent_id: string | null;
  /** 来源模板节点，null = 手工新建（不进裁剪表） */
  template_id: string | null;
  /** 来源裁剪表，null = 手工新建。多张表都保留同一个评审时，靠它区分同名行 */
  tailoring_id: string | null;
  tailoring_title: string | null;
  kind: EStageReviewKind;
  title: string;
  status: EStageReviewStatus;
  /** 空串 = 还没有结论 */
  result: EStageReviewResult | "";
  leader_id: string | null;
  leader_detail: IUserLite | null;
  auditor_id: string | null;
  auditor_detail: IUserLite | null;
  start_date: string | null;
  end_date: string | null;
  attachment_count: number;
  comment_count: number;
  is_manual: boolean;
  sort_order: number;
  created_at: string;
  updated_at: string;
};

export type TStageReviewDetail = TStageReview & {
  stage_detail: TDataDictionaryItemLite | null;
  /** 评审活动才有：所属评审的标题，只用来画面包屑 */
  parent_title: string | null;
  description_html: string | null;
  work_instruction: string;
  conditional_reason: string;
  /** 模板上快照下来的角色名称文本，负责人候选人按它从产品成员里筛 */
  initiator_role: string;
  leader_role: string;
  auditor_role: string;
  production_mode: TProductionMode | "";
  shipment_assessment: TShipmentAssessment | "";
  /** 成品：只有「O阶段评审」有值 */
  finished_goods: {
    akf_code: string;
    production_quantity: number | null;
    product_config: string;
    baseline_archive_code: string;
    components: string[];
  };
  /** 组件版本：同样只有「O阶段评审」有值，这一组只有「版本」一个子属性 */
  component_versions: { version: string };
  created_by: string | null;
};

/** 左栏的一个阶段：只列出真的有评审的阶段，四个状态各自的条数给分段进度条用 */
export type TStageReviewStageSummary = {
  stage_id: string;
  label: string;
  total: number;
  /** 只有产品级汇总有：这个阶段的评审分布在几个项目里 */
  project_count?: number;
  not_started: number;
  in_review: number;
  in_approval: number;
  completed: number;
};

/**
 * 产品页的阶段评审：关联项目里这个产品的全部评审（只含当前用户能看的项目）。
 * `linked_project_count` 为 0 时是「还没有关联项目」，而不是「关联了但还没生成评审」。
 */
export type TProductStageReviewsResponse = {
  linked_project_count: number;
  linked_project_ids: string[];
  reviews: TStageReview[];
};

export type TStageReviewComment = {
  id: string;
  workspace: string;
  project: string;
  stage_review: string;
  actor: string | null;
  actor_detail: IUserLite | null;
  comment_stripped: string;
  comment_json: unknown;
  comment_html: string;
  parent: string | null;
  edited_at: string | null;
  created_at: string;
  updated_at: string;
  created_by: string | null;
  updated_by: string | null;
};

export type TStageReviewActivity = {
  id: string;
  stage_review: string;
  actor: string | null;
  actor_detail: IUserLite | null;
  verb: string;
  field: string | null;
  old_value: string | null;
  new_value: string | null;
  /** 负责人 / 审核者这类成员字段：值是显示名，identifier 是用户 id；老记录为空 */
  old_identifier: string | null;
  new_identifier: string | null;
  comment: string;
  stage_review_comment: string | null;
  extra: Record<string, unknown>;
  created_at: string;
};

export type TStageReviewAttachment = {
  id: string;
  name: string;
  size: number;
  type: string;
  is_uploaded: boolean;
  created_at: string;
  created_by_id: string | null;
  created_by_detail: IUserLite | null;
};

/**
 * 负责人 / 审核者的候选人。`source` 是这批人从哪一层找出来的：
 * `product` 该产品下担任这个角色的人 → `workspace` 工作区里担任这个角色的人 →
 * `project` 都没有，回落到全部项目成员。
 */
export type TStageReviewCandidateSource = "product" | "workspace" | "project";

export type TStageReviewCandidates = {
  role_name: string;
  source: TStageReviewCandidateSource;
  results: IUserLite[];
};

export type TCreateStageReviewPayload = {
  product_id: string;
  stage_id: string;
  kind: EStageReviewKind;
  parent_id?: string | null;
  title: string;
  description_html?: string;
  work_instruction?: string;
  leader_id?: string | null;
  auditor_id?: string | null;
  start_date?: string | null;
  end_date?: string | null;
};

/** 详情里能改的字段。**status 与 result 不在其中** —— 那两列只能由动作推进 */
export type TUpdateStageReviewPayload = Partial<{
  title: string;
  description_html: string;
  work_instruction: string;
  leader: string | null;
  auditor: string | null;
  start_date: string | null;
  end_date: string | null;
  akf_code: string;
  production_quantity: number | null;
  product_config: string;
  baseline_archive_code: string;
  components: string[];
  component_version: string;
}>;

/**
 * 提交审核时的结论。除「通过」外都必须带结论说明（conditional_reason），O 阶段必须带
 * 生产方式与出货评估。落点由结论决定：不通过留在评审中，其余进审核中。
 */
/**
 * 退回上一步的理由，必填。它只进轨迹（`extra.rollback_reason`），不落在评审字段上 ——
 * 一条评审可以被退回多次，每次的理由都属于那一次事件。与结论说明不是同一项。
 */
export type TRollbackStageReviewPayload = {
  reason: string;
};

export type TSubmitStageReviewPayload = {
  result: EStageReviewResult;
  conditional_reason?: string;
  production_mode?: TProductionMode | "";
  shipment_assessment?: TShipmentAssessment | "";
};
